File: boxlid.js
// a jquery plugin for a box lid type layout
(function($) {

	$.fn.boxlid = function(method) {
		var methods = {
			init: function(options) {
				var that = this;
				// initialize box lid layout

				// initialize stored data if it doesn't exist
				if(!that.data("boxlid")) {
					that.data("boxlid", {callbacks: {}});
				}

				// wrapper init
				this.addClass("boxlid-wrapper");

				// top panel init:
				// add the handle
				$(".boxlid-top-panel", this).append(
					$("<div></div>", {'class': 'boxlid-top-handle'})
				);

				// right panel:
				// add handle
				$(".boxlid-right-panel", this).append(
					$("<div></div>", {'class': 'boxlid-right-handle'})
				);

				// left panel:
				// add handle
				$(".boxlid-left-panel", this).append(
					$("<div></div>", {'class': 'boxlid-left-handle'})
				);

				// bottom panel:
				// add handle
				$(".boxlid-bottom-panel", this).prepend(
					$("<div></div>", {'class': 'boxlid-bottom-handle'})
				);

				this.boxlid("resize");

				// add interaction handlers
				var toppanel = $(".boxlid-top-panel", this);
				var leftpanel = $(".boxlid-left-panel", this);
				var rightpanel = $(".boxlid-right-panel", this);
				var bottompanel = $(".boxlid-bottom-panel", this);

				// top handle drag
				$(".boxlid-top-handle", that).mousedown(function(event) {
					$("body").on("mousemove.boxlid", function(event) {

						// resize to set top height
						that.boxlid('resize', {top: {height: event.pageY - toppanel.offset().top}});

						return false;
					});
				});
				// left handle drag
				$(".boxlid-left-handle", that).mousedown(function(event) {
					$("body").on("mousemove.boxlid", function(event) {

						// resize to set left width
						that.boxlid('resize', {left: {width: event.pageX - leftpanel.offset().left}});

						return false;
					});
				});
				// right handle drag
				$(".boxlid-right-handle", that).mousedown(function(event) {
					$("body").on("mousemove.boxlid", function(event) {

						// resize to set right width
						that.boxlid('resize', {right: {width: rightpanel.offset().left + rightpanel.width() - event.pageX}});

						return false;
					});
				});
				// bottom handle drag
				$(".boxlid-bottom-handle", that).mousedown(function(event) {
					$("body").on("mousemove.boxlid", function(event) {
						// resize to set bottom height
						that.boxlid('resize', {bottom: {height: bottompanel.offset().top + bottompanel.height() - event.pageY}});

						return false;
					})
				})

				// up handlers
				$("body").mouseup(function(event) {
					$("body").off("mousemove.boxlid");
					log.log('taking mouse move off, because up', 'handle');
				});

				// resize handler
				this.on('resize.boxlid', function(event) {
					log.log('resizing', 'handle');
					this.boxlid('resize');
				});

				// register callbacks if they exist
				if(options && options.callbacks) {
					this.boxlid('callback', options.callbacks);
				}

				// resize to supplied sizes if they exist
				if(options && options.sizes) {
					this.boxlid('resize', options.sizes);
				}
			},
			// set the sizes of one or more of the panels by any css or pixel integer, and
			// pixelize the resulting sizes and fill the container area
			// options = {side: {dimension: size, +}, +}
			// where	side = (top | right | left | bottom)
			//			dimension = (width | height)
			resize: function(options) {
				log.log('resizing', 'handle');
				var that = this;

				// set sizes from options if supplied
				var opposites = {top: {height: 'left',
								 		width: 'right'},
								right: {height: 'bottom',
										width: 'top'},
								left: {height: 'top',
										width: 'bottom'},
								bottom: {height: 'right',
										width: 'left'}
								};
				if(options) {
					$.each(options, function(side, val) {
						$.each(val, function(dimension, size) {
							// set css width or height
							// allow css string or integer number of pixels
							$(".boxlid-" + side + "-panel", that).css(dimension, typeof size === 'string' ? size : size + 'px');
							// call callback if it exists
							if(that.data('boxlid').callbacks[side]) {
								that.data('boxlid').callbacks[side]($(".boxlid-" + side + "-panel", that)[dimension]() - 1);
							}
							// set width or height of opposite side
							$(".boxlid-" + opposites[side][dimension] + "-panel", that)
								.css(dimension, that[dimension]() - $(".boxlid-" + side + "-panel", that)[dimension]() - 1);
							// call callback for adjacent panel if callback exists
							if(that.data('boxlid').callbacks[opposites[side][dimension]]) {
								that.data('boxlid').callbacks[opposites[side][dimension]](
									$(".boxlid-" + opposites[side][dimension] + "panel", that)[dimension]()
								);
							}
						});
					});
					// resize center div
					$(".boxlid-center", that).width(
						$(".boxlid-right-panel", that).offset().left - $(".boxlid-left-panel", that).offset().left -
						$(".boxlid-left-panel", that).outerWidth()
					);
					$(".boxlid-center", that).height(
						$(".boxlid-bottom-panel", that).offset().top - $(".boxlid-top-panel", that).offset().top -
						$(".boxlid-top-panel", that).outerHeight()
					);
					// call center resize callback if it exists
					if(that.data('boxlid').callbacks.center) {
						that.data('boxlid').callbacks.center({width: $(".boxlid-center", that).width(),
																height: $(".boxlid-center", that).height()});
					}
				}
				this.boxlid('pixelizeSizes');
			},
			// force all sizes to be set in pixels and to fill the area of the container
			pixelizeSizes: function() {
				// force all to be in pixels

				// set sizes in pixels
				// set width in pixels for top and left
				$(".boxlid-top-panel, .boxlid-left-panel", this).width(function(index, width) {
					return width;
				});
				// set width in pixels for right and bottom
				$(".boxlid-right-panel", this).width(this.width() - $(".boxlid-top-panel").width() - 1);
				$(".boxlid-bottom-panel", this).width(this.width() - $(".boxlid-left-panel").width() - 1);

				// set height in pixels for top and right
				$(".boxlid-top-panel, .boxlid-right-panel", this).height(function(index, height) {
					return height;
				});
				// set height in pixels for left and bottom
				$(".boxlid-left-panel", this).height(this.height() - $(".boxlid-top-panel").height() - 1);
				$(".boxlid-bottom-panel", this).height(this.height() - $(".boxlid-right-panel").height() - 1);
			},
			// set callbacks for when a panel is resized
			callback: function(callbackinfo) {
				var that = this;
				$.each(callbackinfo, function(panel, callback) {
					// store callback
					that.data('boxlid').callbacks[panel] = callback;
				});
			}
		};

		if(methods[method]) {
			return methods[method].apply(this, Array.prototype.slice.call(arguments, 1));
		} else if(typeof method == 'object' || ! method) {
			return methods.init.apply(this, arguments);
		} else {
			$.error('Method ' + method + ' does not exist on jQuery.cake');
		}

	    return this;
	}

})(jQuery);

File: dart/BoxLid.dart
/** [BoxLid] is a class to encapsulate the functionality
 * of the box-lid layout
 **/
#library("BoxLid");
#import("dart:html");

class BoxLid {
  // the root element of the boxlid layout
  Element root;
 
  // convenience accessors to panels
  Element get _leftPanel() => root.query(".boxlid-left-panel");
  Element get _rightPanel() => root.query(".boxlid-right-panel");
  Element get _topPanel() => root.query(".boxlid-top-panel");
  Element get _bottomPanel() => root.query(".boxlid-bottom-panel");
  Element get _centerPanel() => root.query(".boxlid-center");

  // the width or height of each panel
  int _left, _right, _top, _bottom;
  // the last width or height of each panel before it was closed
  int _openLeft, _openRight, _openTop, _openBottom;
  // the width or height of the panels when closed;
  int _closedSize;
  // mouse move handler storage
  Map<String, Function> _moveHandlers;

  // accessors for the widths and heights
  // setters automatically update dependent sizes
  int get left() => _left;
  int set left (value) {
    _left = value;
    // set sizes of elements
    _leftPanel.style.width = "${value}px";
    _bottomPanel.style.width = "${root.$dom_offsetWidth - value - 1}px";
    _setCenterSize();
  }
  int get top() => _top;
  int set top(value) {
    _top = value;
    // set sizes of elements
    _topPanel.style.height = "${value}px";
    _leftPanel.style.height = "${root.$dom_offsetWidth - value - 1}px";
    _setCenterSize();
  }
  int get bottom() => _botttom;
  int set bottom(value) {
    _bottom = value;
    // set sizes of elements
    _bottomPanel.style.height = "${value}px";
    _rightPanel.style.height = "${root.$dom_offsetHeight - value - 1}px";
    _setCenterSize();
  }
  int get right() => _right;
  int set right(value) {
    _right = value;
    // set sizes of elements
    _rightPanel.style.width = "${value}px";
    _topPanel.style.width = "${root.$dom_offsetWidth - value - 1}px";
    _setCenterSize();
  }

  /** Set the size of the center panel from the sizes of the side panels **/
  _setCenterSize() {
    int width = root.$dom_offsetWidth -
                  _leftPanel.$dom_offsetWidth -
                  _rightPanel.$dom_offsetWidth;
    int height = root.$dom_offsetHeight -
                  _topPanel.$dom_offsetHeight -
                  _bottomPanel.$dom_offsetHeight;
    _centerPanel.style.width = "${width}px";
    _centerPanel.style.height = "${height}px";
  }

  BoxLid(Element element) {
    _closedSize = 10;
    root = element;
    // add wrapper class
    root.classes.add("boxlid-wrapper");

    // top panel init:
    // add the handle
    Element topHandle = new Element.tag("div");
    topHandle.classes.add("boxlid-top-handle");
    root.query(".boxlid-top-panel").nodes.add(topHandle);

    // right panel:
    // add handle
    Element rightHandle = new Element.tag("div");
    rightHandle.classes.add("boxlid-right-handle");
    root.query(".boxlid-right-panel").nodes.add(rightHandle);

    // left panel:
    // add handle
    Element leftHandle = new Element.tag("div");
    leftHandle.classes.add("boxlid-left-handle");
    root.query(".boxlid-left-panel").nodes.add(leftHandle);

    // bottom panel
    // add handle
    Element bottomHandle = new Element.tag("div");
    bottomHandle.classes.add("boxlid-bottom-handle");
    _bottomPanel.insertBefore(bottomHandle, _bottomPanel.nodes[0]);

    // update sizes
    //this.resize();
    left = _leftPanel.$dom_offsetWidth;
    right = _rightPanel.$dom_offsetWidth;

    // add interaction handlers
    // store move handler
    _moveHandlers = {
      "top": (event) => top = event.pageY - _topPanel.$dom_offsetTop,
      "left": (event) => left = event.pageX - _leftPanel.$dom_offsetLeft,
      "right": (event) => right = _rightPanel.offset().left + _rightPanel.width - event.pageX,
      "bottom": (event) => bottom = _bottomPanel.offset().top + _bottomPanel.height - event.pageY
    };
    topHandle.on.mouseDown.add((event) {
      // add move event
      root.on.mouseMove.add(_moveHandlers["top"]);
    });
    leftHandle.on.mouseDown.add((event) {
      root.on.mouseMove.add(_moveHandlers["left"]);
    });
    rightHandle.on.mouseDown.add((event) {
      root.on.mouseMove.add(_moveHandlers["right"]);
    });
    bottomHandle.on.mouseDown.add((event) {
      root.on.mouseMove.add(_moveHandlers["bottom"]);
    });
    
    // remove handlers on up
    root.on.mouseUp.add((event) {
      // remove move handler
      //root.on.mouseMove.remove(_moveHandlers["top"]);
      for(var handler in _moveHandlers.getValues()) {
        root.on.mouseMove.remove(handler);
      }
    });

    // click handles to open / close them
    leftHandle.on.click.add((event) {
      //if(_leftPanel.)
    });

  }
}